modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/pipelining/HttpPipeliningHandler.java

```diff
@@ -120,4 +121,13 @@ public void write(final ChannelHandlerContext ctx, final Object msg, final Chann
         }
     }
 
+    @Override
+    public void close(ChannelHandlerContext ctx, ChannelPromise promise) throws Exception {
+        ClosedChannelException closedChannelException = new ClosedChannelException();
```

Comment:
I don't think the synchronization is necessary (I think it can actually all be removed from this class) but let's put that aside for now as a separate issue.

I meant:

```diff
diff --git a/modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/pipelining/HttpPipeliningHandler.java b/modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/pipelining/HttpPipeliningHandler.java
index c09099a221..e9d99378e2 100644
--- a/modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/pipelining/HttpPipeliningHandler.java
+++ b/modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/pipelining/HttpPipeliningHandler.java
@@ -123,10 +123,12 @@ public class HttpPipeliningHandler extends ChannelDuplexHandler {
 
     @Override
     public void close(ChannelHandlerContext ctx, ChannelPromise promise) throws Exception {
-        ClosedChannelException closedChannelException = new ClosedChannelException();
-        for (HttpPipelinedResponse pipelinedResponse : holdingQueue) {
-            pipelinedResponse.release();
-            pipelinedResponse.promise().setFailure(closedChannelException);
+        if (!holdingQueue.isEmpty()) {
+            final ClosedChannelException closedChannelException = new ClosedChannelException();
+            for (final HttpPipelinedResponse pipelinedResponse : holdingQueue) {
+                pipelinedResponse.release();
+                pipelinedResponse.promise().setFailure(closedChannelException);
+            }
         }
         ctx.close(promise);
     }
```

(and sure, we can add the synchronized block for now until I get around to investigating removing it).